chamber/models/__init__.py

```diff
@@ -41,52 +43,74 @@ def model_to_dict(instance, fields=None, exclude=None):
-    """
-    A model mixin that tracks model fields' values and provide some useful api
-    to know what fields have been changed.
-    """
+ValueChange = collections.namedtuple('ValueChange', 'initial current')
 
-    def __init__(self, *args, **kwargs):
-        super(ModelDiffMixin, self).__init__(*args, **kwargs)
-        self.__initial = self._dict
 
-    @property
-    def initial_values(self):
-        return self.__initial
+class ChangedFields(object):
+
+    def __init__(self, instance):
+        self.instance = instance
+        self.initial_values = self.get_instance_dict(instance)
+
+    def get_instance_dict(self, instance):
+        return model_to_dict(instance, fields=(field.name for field in instance._meta.fields))
 
     @property
     def diff(self):
-        d1 = self.__initial
-        d2 = self._dict
-        diffs = [(k, (v, d2[k])) for k, v in d1.items() if v != d2[k]]
-        return dict(diffs)
+        d1 = self.initial_values
+        d2 = self.get_instance_dict(self.instance)
+        return {k : ValueChange(v, d2[k]) for k, v in d1.items() if v != d2[k]}
 
-    @property
-    def has_changed(self):
-        return bool(self.diff)
+    def __setitem__(self, key, item):
+        raise AttributeError('Object is readonly')
 
-    @property
-    def changed_fields(self):
-        return set(self.diff.keys())
+    def __getitem__(self, key):
+        return self.diff[key]
 
-    def get_field_diff(self, field_name):
-        """
-        Returns a diff for field if it's changed and None otherwise.
-        """
-        return self.diff.get(field_name, None)
+    def __repr__(self):
+        return repr(self.diff)
 
-    def save(self, *args, **kwargs):
-        """
-        Saves model and set initial state.
-        """
-        super(ModelDiffMixin, self).save(*args, **kwargs)
-        self.__initial = self._dict
+    def __len__(self):
+        return len(self.diff)
 
-    @property
-    def _dict(self):
-        return model_to_dict(self, fields=[field.name for field in
-                             self._meta.fields])
+    def __delitem__(self, key):
+        raise AttributeError('Object is readonly')
+
+    def clear(self):
+        raise AttributeError('Object is readonly')
+
+    def has_key(self, k):
+        return self.diff.has_key(k)
+
+    def has_any_key(self, *keys):
+        return bool(set(self.keys()) & set(keys))
+
+    def update(self, *args, **kwargs):
+        raise AttributeError('Object is readonly')
+
+    def keys(self):
+        return self.diff.keys()
+
+    def values(self):
+        return self.diff.values()
+
+    def items(self):
+        return self.diff.items()
+
+    def pop(self, *args, **kwargs):
+        raise AttributeError('Object is readonly')
+
+    def __cmp__(self, dict):
+        return cmp(self.diff, dict)
+
+    def __contains__(self, item):
+        return item in self.diff
+
+    def __iter__(self):
+        return iter(self.diff)
+
+    def __unicode__(self):
```

Comment:
teoreticky je tu to unicode jen navíc, jelikož se stejně volá spíš repr. Nicméně máš pravdu přidám anotaci